### ISSUE TITLE
ADD ML MODELS

### ISSUE THREAD

**Issue by fxgeniusllc-oss** [fxgeniusllc-oss/APEX-ARBITRAGE-SYSTEM#2]
# === Transformer-Based Market Maker Model Trainer ===
import os
import torch
import torch.nn as nn
import pandas as pd
import numpy as np
from torch.utils.data import Dataset, DataLoader
from web3 import Web3

# === Blockchain Setup ===
RPC_URL = os.getenv("ALCHEMY_RPC_URL", "https://polygon-mainnet.g.alchemy.com/v2/YXw_o8m9DTfqafsqX3ebqH5QP1kClfZG")
w3 = Web3(Web3.HTTPProvider(RPC_URL))

# === Data Loader ===
def fetch_block_data(block_number):
    block = w3.eth.get_block(block_number, full_transactions=True)
    return {
        "timestamp": block.timestamp,
        "tx_count": len(block.transactions),
        "gas_used": block.gasUsed,
        "value_transfer": sum([tx.value for tx in block.transactions]) / 1e18
    }

def build_dataset(start_block, end_block):
    return pd.DataFrame([fetch_block_data(b) for b in range(start_block, end_block)])

# === Dataset ===
class SequenceDataset(Dataset):
    def __init__(self, df, seq_len):
        features = df.drop("timestamp", axis=1).values
        self.X = np.array([features[i:i+seq_len] for i in range(len(features) - seq_len)], dtype=np.float32)
        self.y = np.array([features[i+seq_len][-1] for i in range(len(features) - seq_len)], dtype=np.float32)
    def __len__(self): return len(self.X)
    def __getitem__(self, idx): return self.X[idx], self.y[idx]

# === Transformer Model ===
class TransformerMarketMaker(nn.Module):
    def __init__(self, input_dim, d_model=128, heads=4, layers=3):
        super().__init__()
        self.project = nn.Linear(input_dim, d_model)
        encoder = nn.TransformerEncoderLayer(d_model, heads)
        self.encoder = nn.TransformerEncoder(encoder, num_layers=layers)
        self.output = nn.Sequential(nn.Linear(d_model, 64), nn.ReLU(), nn.Linear(64, 1))
    def forward(self, x):
        x = self.project(x).permute(1,0,2)
        return self.output(self.encoder(x)[-1])

# === Training Function ===
def train_market_maker(start_block, end_block, model_path="./models/transformer_market_maker.pth"):
    df = build_dataset(start_block, end_block)
    ds = SequenceDataset(df, seq_len=64)
    dl = DataLoader(ds, batch_size=128, shuffle=True)
    model = TransformerMarketMaker(input_dim=ds.X.shape[2]).cuda()
    opt = torch.optim.Adam(model.parameters(), lr=1e-4)
    loss_fn = nn.MSELoss()

    for epoch in range(10):
        total = 0
        for xb, yb in dl:
            xb, yb = xb.cuda(), yb.cuda().unsqueeze(1)
            opt.zero_grad(); loss = loss_fn(model(xb), yb); loss.backward(); opt.step()
            total += loss.item() * xb.size(0)
        print(f"Epoch {epoch+1}: Loss = {total / len(dl.dataset):.6f}")

    torch.save(model.state_dict(), model_path)
    print(f"✅ Saved model to {model_path}")

if __name__ == "__main__":
    train_market_maker(46000000, 46000100)

[risk_analyzer_model.py](https://github.com/user-attachments/files/23031337/risk_analyzer_model.py)
[market_maker_model.py](https://github.com/user-attachments/files/23031338/market_maker_model.py)
[profit_predictor_model.py](https://github.com/user-attachments/files/23031339/profit_predictor_model.py)

[market_maker_model.py](https://github.com/user-attachments/files/23031340/market_maker_model.py)
[profit_predictor_model.py](https://github.com/user-attachments/files/23031342/profit_predictor_model.py)
[risk_analyzer_model.py](https://github.com/user-attachments/files/23031341/risk_analyzer_model.py)
